components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java

```diff
@@ -267,16 +327,124 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
             }
         }
 
+        // Handle the IDP mapped user roles claim and app role claim.
+        if (useAppAssociatedRoles) {
+            if (CollectionUtils.isNotEmpty(federatedUserRolesUnmappedExclusive)) {
+                // Adding identity provider mapped user roles to be used in the federated user role resolver
+                // for scope validation.
+                spFilteredClaims.put(FrameworkConstants.IDP_MAPPED_USER_ROLES,
+                        String.join(FrameworkUtils.getMultiAttributeSeparator(), federatedUserRolesUnmappedExclusive));
+                spFilteredClaims.put(FrameworkConstants.APP_ROLES_CLAIM,
+                        String.join(FrameworkUtils.getMultiAttributeSeparator(),
+                                federatedUserRolesUnmappedExclusive));
+            } else {
+                // If current authenticator is OrganizationAuthenticator, then get the roles claim from
+                // the user attributes, and set it to the identity provider mapped user roles claim.
+                if (stepConfig.getAuthenticatedAutenticator().getApplicationAuthenticator()
+                        .getName().equals(FrameworkConstants.ORGANIZATION_AUTHENTICATOR)) {
+                    // Get roles claim from the user attributes.
+                    String rolesClaim = remoteClaims.get(localToIdPClaimMap.get(FrameworkConstants.ROLES_CLAIM));
+                    spFilteredClaims.put(FrameworkConstants.IDP_MAPPED_USER_ROLES, rolesClaim);
+                } else {
+                    spFilteredClaims.put(FrameworkConstants.IDP_MAPPED_USER_ROLES, StringUtils.EMPTY);
+                }
+            }
+        }
 
         //Add multi Attributes separator with claims.it can be defined in user-mgt.xml file
         UserRealm realm = getUserRealm(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
         UserStoreManager userStore = getUserStoreManager(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, realm);
         addMultiAttributeSeparatorToRequestedClaims(null, userStore, spFilteredClaims, realm);
 
         return spFilteredClaims;
+    }
+
+    /**
+     * Map the external IDP roles to local roles.
+     * If excludeUnmapped is true exclude unmapped roles.
+     * Otherwise include unmapped roles as well.
+     *
+     * @param externalIdPConfig     External IDP config.
+     * @param extAttributesValueMap External IDP attributes.
+     * @param idpRoleClaimUri       IDP role claim URI.
+     * @param excludeUnmapped       Whether to exclude unmapped roles.
+     * @return ArrayList<string> List of identity provider mapped user roles.
+     */
```

Comment:
exception is not documented in the comment

Reply:
Fixed with